src/tools.ts

```diff
@@ -28,6 +17,123 @@ function getRangesFromDecorations (
     .map((decoration) => decoration.range)
 }
 
+function minusRanges (uniqueRange: monaco.Range, ranges: monaco.Range[]): monaco.Range[] {
+  const newRanges: monaco.Range[] = []
+  let lastEndLineNumber = uniqueRange.startLineNumber
+  let lastEndColumn = uniqueRange.startColumn
+
+  for (const range of ranges) {
+    const newRange = new monaco.Range(lastEndLineNumber, lastEndColumn, range.startLineNumber, range.startColumn)
+    lastEndLineNumber = range.endLineNumber
+    lastEndColumn = range.endColumn
+    newRanges.push(newRange)
+  }
+
+  if (lastEndLineNumber < uniqueRange.endLineNumber || lastEndColumn < uniqueRange.endColumn) {
+    newRanges.push(new monaco.Range(lastEndLineNumber, lastEndColumn, uniqueRange.endLineNumber, uniqueRange.endColumn))
+  }
+
+  return newRanges
+}
+
+function createNewOperation (
+  oldOperation: ValidAnnotatedEditOperation,
+  newRange: monaco.Range,
+  newText: string,
+  index: number
+): ValidAnnotatedEditOperation {
+  const identifier = oldOperation.identifier != null
+    ? { major: oldOperation.identifier.major, minor: oldOperation.identifier.minor + index }
+    : null
+  return new ValidAnnotatedEditOperation(
+    identifier,
+    newRange,
+    newText,
+    oldOperation.forceMoveMarkers,
+    oldOperation.isAutoWhitespaceEdit,
+    oldOperation._isTracked
+  )
+}
+
+function computeNewOperationsForLockedCode (
+  editor: monaco.editor.ICodeEditor,
+  decorationFilter: (decoration: monaco.editor.IModelDecoration) => boolean,
+  editorOperations: ValidAnnotatedEditOperation[],
+  withDecoration: boolean,
+  displayLockedCodeError: (position: monaco.Position) => void
+): ValidAnnotatedEditOperation[] {
+  const model = editor.getModel()
+  if (model == null) {
+    return []
+  }
+
+  const fullModelRange = model.getFullModelRange()
+  const ranges = getRangesFromDecorations(editor, decorationFilter)
+  const uneditableRanges = withDecoration ? ranges : minusRanges(fullModelRange, ranges)
+  if (uneditableRanges.length <= 0) {
+    return editorOperations
+  }
+
+  const newOperations: ValidAnnotatedEditOperation[] = []
+  for (const operation of editorOperations) {
+    const operationRange = operation.range
+    const uneditableRangesThatIntersects = uneditableRanges.filter(range => monaco.Range.areIntersecting(range, operationRange))
+
+    // The operation range doesn't intersect with an uneditable range
+    if (uneditableRangesThatIntersects.length <= 0) {
+      newOperations.push(operation)
+      continue
+    }
+
+    // The operation range is entirely in an uneditable range
+    if (uneditableRangesThatIntersects.some(range => range.containsRange(operationRange))) {
+      displayLockedCodeError(
+        new monaco.Position(operationRange.startLineNumber, operationRange.startColumn))
```

Comment:
It's probably not the responsability of this fonction to display an error

Reply:
I removed the call to `displayLockedCodeError` while refactoring the function: b598071.